Add rules for controller-card power-off (#1041)
openconfig/public: doc/controller-card_poweroff.md

## Objective:

This document introduces the concept of persistent shutdown of controller-cards in switches/routers with dual controller-cards, outlines rules to determine its use and proposes new telemetry counters to support the feature.

> ## Possible use cases for the gNMI approach:

Reviewer: The openconfig yang is not gnmi specific (ie: many people use netconf for example)

1. Operator suspects a network device issue is occurring due to faults on a CONTROLLER_CARD. The operator wants to quickly disable the CONTROLLER_CARD and later, troubleshoot the issue in a maintenance window. The operator also wants to ensure that the card stays disabled even if the device is rebooted. Some example scenarios below:
   

>    a. It is possible that the faulty controller card is in a boot loop and the operator finds it best to disable the card to prevent undesired state. Expectations are also that the state persists even post a system reboot.

Reviewer: ```suggestion
   a. It is possible that the faulty controller card is in a boot loop and the operator finds it best to disable the card to prevent undesired state. Expectations are also that the power remains disabled  even after a system reboot.
```

> 2. It is possible that the faulty controller card is in a boot loop and the operator finds it best to disable the card to prevent undesired state. Expectations are also that the state persists even post a system reboot.

Reviewer: This reads as an example of item 1.  Indent this?  Maybe add another example where a CPU is overheating if left powered on for more than a few minutes.

Author: Done, ptal!

>    b. CPU overheating, memory errors or hardware problems with certain components of the PRIMARY card may require the operator to proactively shutdown the card and not let it be online during regular operation until RMA is completed.

Reviewer: ```suggestion
   b. CPU overheating, memory errors or other hardware problems with the PRIMARY card may require the operator to proactively shutdown the card and not let it be online during regular operation until replacement of the hardware is completed.
```

Author: done

> 4. Following an RMA, if a new issue arises or the original problem remains, operators may instruct field techs to leave the replacement card installed. This allows for replacement upon arrival or controlled troubleshooting. In these cases, the ability to control the card's operational state is advantageous.

Author: removed #4

> ## Proposed Rules

Reviewer: ```suggestion
## CONTROLLER_CARD power and redundancy requirements 
```

> Idea here is that the subject controller-card can be disabled using the OC-Path [/components/component/controller-card/config/power-admin-state](https://openconfig.net/projects/models/schemadocs/yangdoc/openconfig-platform.html#components-component-controller-card-config-power-admin-state) by setting its value to POWER\_DISABLED. Following rules:

Reviewer: ```suggestion
Electrical power for a CONTROLLER_CARD can be turned off using the OC-Path [/components/component/controller-card/config/power-admin-state](https://openconfig.net/projects/models/schemadocs/yangdoc/openconfig-platform.html#components-component-controller-card-config-power-admin-state) and setting its value to POWER\_DISABLED.  The following rules regarding CONTROLLER_CARD redundancy and changes to `power-admin-state` should be followed by the device:
```

> Electrical power for a CONTROLLER_CARD can be turned off using the OC-Path [/components/component/controller-card/config/power-admin-state](https://openconfig.net/projects/models/schemadocs/yangdoc/openconfig-platform.html#components-component-controller-card-config-power-admin-state) and setting its value to POWER\_DISABLED.  The following rules regarding CONTROLLER_CARD redundancy and changes to `power-admin-state` should be followed by the device:

Reviewer: ```suggestion
Electrical power for a CONTROLLER_CARD can be configured off using the OC-Path [/components/component/controller-card/config/power-admin-state](https://openconfig.net/projects/models/schemadocs/yangdoc/openconfig-platform.html#components-component-controller-card-config-power-admin-state) and setting its value to POWER\_DISABLED.  The following rules regarding CONTROLLER_CARD redundancy and changes to `power-admin-state` should be followed by the device:
```

> 1. Only a CONTROLLER_CARD in state redundant-roleSECONDARY will honor config/power-admin-state set toPOWER_DISABLED. If the controller-card's'redundant-role' is 'PRIMARY', and it's 'config/power-admin-state' is set to 'POWER_DISABLED', the NOS must allow the configuration. However, the change must take effect only on the next reboot. Following scenarios in context

Reviewer: ```suggestion
1. Only a CONTROLLER_CARD in `state/redundant-role` 'SECONDARY' will honor a change in `config/power-admin-state` to POWER_DISABLED. If the controller-card's `redundant-role` is 'PRIMARY', and its `config/power-admin-state` is set to 'POWER_DISABLED', the NOS must allow the configuration. However, the `state/power-admin-state` should remain as 'POWER_ENABLED'.  A change in `state/power-admin-state` must take effect only on the next reboot or if the CONTROLLER_CARD becomes 'SECONDARY'. Examples of scenarios include:
```

> `Scenario#1:`

Reviewer: ```suggestion
### Scenario 1 - Power off a secondary CONTROLLER_CARD
```

> Let's say controller-card0 is PRIMARY and controller-card1 is SECONDARY and controller-card1 receives an operator driven change of config/power-admin-state = POWER_DISABLED, then controller-card0 will power-off controller-card1 immediately.  The leaf state/power-admin-state for controller-card1 must also be POWER_DISABLED. The 'state/last-poweroff-reason/trigger' should show as USER_INITIATED. The NOS may optionally update /state/last-poweroff-reason/details. state/last-poweroff-time should record the time when the card was powered-off. For example:

Reviewer: ```suggestion
Let's say controller-card0 is PRIMARY and controller-card1 is SECONDARY and controller-card1 receives an operator driven change of config/power-admin-state = POWER_DISABLED, then controller-card0 will power-off controller-card1 immediately.  The leaf state/power-admin-state for controller-card1 must also be POWER_DISABLED. The 'state/last-poweroff-reason/trigger' should show as USER_INITIATED. The NOS may optionally update state/last-poweroff-reason/details. state/last-poweroff-time should record the time when the card was powered-off. For example:
```

```
/components/component[name=controller0]/state/redundant-role, PRIMARY
/components/component[name=controller0]/controller-card/config/power-admin-state, POWER_ENABLED
/components/component[name=controller0]/controller-card/state/power-admin-state, POWER_ENABLED
/components/component[name=controller1]/state/redundant-role, SECONDARY
/components/component[name=controller1]/controller-card/config/power-admin-state, POWER_DISABLED
/components/component[name=controller1]/controller-card/state/power-admin-state, POWER_DISABLED
/components/component[name=controller1]/state/last-poweroff-reason/trigger, USER_INITIATED
/components/component[name=controller1]/state/last-poweroff-reason/details, "User initiated Shutdown"
/components/component[name=controller1]/state/last-poweroff-time, 1706590714681765937
```

> `Scenario#2:`

Reviewer: ```suggestion
### Scenario 2 - Power off primary CONTROLLER_CARD
```

> If controller-card0 is PRIMARY and controller-card1 is SECONDARY and if controller-card0 is set to config/power-admin-state = POWER_DISABLED by an operator, then controller-card0 will stay powered-on until the next reboot. The leaf state/power-admin-state must show as POWER_ENABLED. The state/last-poweroff-reason/trigger should show as USER_INITIATED. Also the NOS can update the leaf /state/last-poweroff-reason/details. The leaf state/last-poweroff-time should record the time when the card powers-off post the next reboot. For example: 

Reviewer: ```suggestion
If controller-card0 is PRIMARY and controller-card1 is SECONDARY and if controller-card0 is set to config/power-admin-state = POWER_DISABLED by an operator, then controller-card0 will stay powered-on until the next reboot. `state/power-admin-state` must show as POWER_ENABLED. If a reboot of the PRIMARY CONTROLLER_CARD occurs,  `state/last-poweroff-time` should record the time when the card powers-off, `state/last-poweroff-reason/trigger` should show as USER_INITIATED and `/state/last-poweroff-reason/details` may be updated. For example: 
```

Reviewer: Trigger and and details should only be updated after a power-off occurs (after a reboot in this scenario)

/components/component[name=controller1]/controller-card/state/power-admin-state, POWER_ENABLED
```

2. A controller-card which is in state/redundant-role=SECONDARY and is config/power-admin-state=POWER_DISABLED should remain powered off, even after a reboot. It is possible that after a reboot, both the controller cards are powered-on. However as soon as the configuration is loaded, the system should power-off the subject controller-card.

3. A CONTROLLER_CARD in redundant-role SECONDARY and state/power-admin-state = POWER_DISABLED cannot transition to redundant-role PRIMARY. If the PRIMARY CONTROLLER_CARD goes down, the device will be offline.

4. When a controller card boots up and loads it’s configuration, power-admin-state may be set to POWER_DISABLED. The CONTROLLER_CARD should then power off and never enter into any controller cardprimary/secondary election process. This also means that an implementation shouldn't start a controller card election process until the configuration is loaded and consumed.

## Concerns and possible failure scenarios

1. If a PRIMARY card malfunctions and ends up in a bootloop, would this approach help?

    **Response:**

>     In this scenario the expectations are that the implementation takes steps to initiate a controller-card switchover operation. Therefore, the standby controller card takes over the PRIMARY role and the system stabilizes allowing for gRPC connections to be established. In this situation if the operator pushes a gNMI configuration to shutdown the SECONDARY card, the PRIMARY card must be able to power-off the SECONDARY controller-card. In this scenario the implementation (depending on their architecture) may also initiate a shutdown of the faulty card from the new PRIMARY card.

Reviewer: ```suggestion
    In this scenario the expectations are that the implementation takes steps to initiate a controller-card switchover operation. Therefore, the standby controller card takes over the PRIMARY role and the system stabilizes allowing for gRPC connections to be established. In this situation if the operator pushes a configuration to shutdown the SECONDARY card, the PRIMARY card must be able to power-off the SECONDARY controller-card. In this scenario the implementation (depending on their architecture) may also initiate a shutdown of the faulty card from the new PRIMARY card.
```

2. Since the operation relies on configuration, it is possible that the failure scenario may kick in before the configuration takes effect post a reboot.

    **Response:**

	Response here is the same as "1" above.

> 3. Both controller-cards are functional and the the secondary controller card is attempted to be shutdown

Reviewer: ```suggestion
3. Both controller-cards are functional and the the secondary controller card is attempted to be powered off.
```

Here and throughout, use a consistent term such as 'power off' and 'power on' for referring to power-admin-state.  Shutdown might be confused with some other configuration.

    **Response:**

    The shutdown can be initiated by pushing the command [/components/component[controller-card#]/controller-card/config/power-admin-state = POWER\_DISABLED](https://openconfig.net/projects/models/schemadocs/yangdoc/openconfig-platform.html#components-component-controller-card-config-power-admin-state) to the box. As a result, the secondary controller card whose [/components/component/state/redundant-role](https://openconfig.net/projects/models/schemadocs/yangdoc/openconfig-platform.html#components-component-state-redundant-role) is SECONDARY

>     will be shutdown and its [/components/component[controller-card#]/controller-card/state/power-admin-state](https://openconfig.net/projects/models/schemadocs/yangdoc/openconfig-platform.html#components-component-controller-card-state-power-admin-state) will change to "POWER\_DISABLED". This configuration is expected to stay sticky in the box and when the device reboots, the subject controller-card is expected to remain Powered-off.

Reviewer: ```suggestion
    will be shutdown and its [/components/component[name="my_secondary_controller-card"]/controller-card/state/power-admin-state](https://openconfig.net/projects/models/schemadocs/yangdoc/openconfig-platform.html#components-component-controller-card-state-power-admin-state) will change to "POWER\_DISABLED".  If the configuration is saved and when the device reboots, the subject controller-card is expected to remain powered off.
```

4. Both controller-cards are functional and the the primary controller card is attempted to be shutdown

    **Response:**

    Follow Rule#1 above

5. Only one controller-card is present and the card is attempted to be shutdown.

    **Response:**

    Since the lone controller-card is PRIMARY, please follow Rule#1 above.

6. Faulty card is shutdown and then replaced with a working card but device still has the config of `/components/component[controller-card#]/controller-card/config/power-admin-state = POWER_DISABLED`

    **Response:**

>     This is as per expectations so inserting a new card into the chassis can happen anytime. However, bringing it online happens only in a controlled environment using a gNMI config push operation of `/components/component[controller-card#]/controller-card/config/power-admin-state = POWER_ENABLED`

Reviewer: ```suggestion
    This is as per expectations so inserting a new card into the chassis can happen anytime. However, bringing it online happens only in a controlled environment using a configuration change to set `/components/component[controller-card#]/controller-card/config/power-admin-state = POWER_ENABLED`
```

>     This is as per expectations so inserting a new card into the chassis can happen anytime. However, bringing it online happens only in a controlled environment using a configuration change to set `/components/component[controller-card#]/controller-card/config/power-admin-state = POWER_ENABLED`

Reviewer: ```suggestion
    The new card should remain powered off.  Power should only be enabled on the new card by a configuration change to set `/components/component[controller-card#]/controller-card/config/power-admin-state = POWER_ENABLED`
```

7. Both controller-cards present during the initial boot operation. However, the controller card that is configured as "POWER\_DISABLED" boots up sooner than the other card and takes over the ACTIVE role.

    **Response:**

>     The controller cards should never take PRIMARY/SECONDARY roles unless the configuration is fully loaded and processed. Hence, once the configuration is loaded, the card configured to be "POWER\_DISABLED" should shutdown. The only exception is when it is the only controller card in the chassis. In which case, the controller card can still load irrespective of its power-admin-state being POWER\_DISABLED. In this situation though, the implementation must send a Syslog message of the severity "Warning" to inform the Operator about the situation.

Reviewer: This exception doesn't apply to this scenario because we've said `Both controller-cards present during the initial boot operation`

However, we could add another scenario below where:

Two cards are present, secondary is powered off.  The primary card is removed and the system goes offline.  If the device is rebooted (for example via externally removing and reapplying power ), what should happen?

Author: `This exception doesn't apply to this scenario because we've said Both controller-cards present during the initial boot operation`

Well, not necessarily for all vendors today config load happens before the PRIMARY/SECONDARY election process. Sequence is indeterministic today between config load operation and election process and hence the scenario here is very well a possibility and is orthogonal to the presence of both cards or not. We are calling out this scenario to ensure expectations are clear.

We can definitely add the scenario you mentioned.

Reviewer: The scenario is defined as two controller cards are present, but the the exception notes is:  `only exception is when it is the only controller card in the chassis. `.  Just a logic issue / should be broken out to a separate scenario IMO.

Author: I added Rule#5. Ptal

Reviewer: LGTM.  The rule reference is a little ambiguous.  There are a couple of things in the document labeled `5.` and it's not clear which one is a rule versus a scenario.  One way to fix this would be to use outline numbered headings like:

1. Item
1.1 Item subheading
2. Another item
2.2 Another subheading

/components/component[name=controller0]/controller-card/state/power-admin-state, POWER_ENABLED
/components/component[name=controller1]/state/redundant-role, SECONDARY
/components/component[name=controller1]/controller-card/state/power-admin-state, POWER_ENABLED
```

After controller-card0 transitions to redundant-role SECONDARY:

```
/components/component[name=controller0]/state/redundant-role, SECONDARY
/components/component[name=controller0]/controller-card/config/power-admin-state, POWER_DISABLED
/components/component[name=controller0]/controller-card/state/power-admin-state, POWER_DISABLED
/components/component[name=controller0]/state/last-poweroff-reason/trigger, USER_INITIATED
/components/component[name=controller0]/state/last-poweroff-reason/details, "User initiated Shutdown"
/components/component[name=controller0]/state/last-poweroff-time, 1706590714681765937
/components/component[name=controller1]/state/redundant-role, PRIMARY
/components/component[name=controller1]/controller-card/config/power-admin-state, POWER_ENABLED
/components/component[name=controller1]/controller-card/state/power-admin-state, POWER_ENABLED
```

> 2. A controller-card which is in state/redundant-role=SECONDARY and is config/power-admin-state=POWER_DISABLED should remain powered off, even after a reboot. It is possible that after a reboot, both the controller cards are powered-on. However as soon as the configuration is loaded, the system should power-off the subject controller-card.

Reviewer 2: Perhaps the term "reboot" is used somewhat ambiguously throughout the text, but here it is stated that the card _should_ remain powered off, but may be powered-on temporarily. And at the end of the document, in the scenario # 8, you provide an example where a reboot (through power cycling) must initiate a new boot cycle.

>If the device is rebooted (for example via externally removing and reapplying power ), what should happen?
>    If during reboot the device continues to have a single controller-card, then Rule#5 above should be followed.

Reviewer: To increase clarity, I recommend replacing 
`It is possible that after a reboot, both the controller cards are powered-on` with 
`During a cold-boot, some implementations may power up both CONTROLLER_CARDS.`

Author: Why only cold boot? We are not asking implementations to maintain state even during warm boot. Therefore it is possible that both controller-cards may turn-on simultaneously post a reboot (cold or warm). Expectations are that the implementation loads the configuration on both cards, the card w/ `power-admin-state=POWER_DISABLED` will be powered-off unless it is the only card.

Reviewer 2: In my view, the ambiguity is that in order to follow the rules that you specified, e.g. the one I quoted:

>If the device is rebooted (for example via externally removing and reapplying power ), what should happen?
If during reboot the device continues to have a single controller-card, then Rule#5 above should be followed.

there must be a way to "reboot" (in your terms) the card that was previously powered-off via the config, and it cannot stay powered off during that time.

But rule 2 says otherwise:
> A controller-card which is in state/redundant-role=SECONDARY and is config/power-admin-state=POWER_DISABLED should remain powered off, even after a reboot.

if the card remains powered off, it will never know if its a single card in the system, or there's another card. So it cannot remain powered off.

Author: As per the offline discussions, @LimeHat seems like the requirement is understood. How an implementation implements the change is left on them.

3. A CONTROLLER_CARD in redundant-role SECONDARY and state/power-admin-state = POWER_DISABLED cannot transition to redundant-role PRIMARY. If the PRIMARY CONTROLLER_CARD goes down, the device will be offline.

> 4. When a controller card boots up and loads it’s configuration, power-admin-state may be set to POWER_DISABLED. The CONTROLLER_CARD should then power off and never enter into any controller card primary/secondary election process. This also means that an implementation shouldn't start a controller card election process until the configuration is loaded and consumed.

Reviewer 2: If both controller cards have been set as POWER_DISABLED, does that mean that both cards should be powered off?

Also, there's a slight contradiction with the next rule (# 5). 
>The CONTROLLER_CARD should then power off and never enter into any controller card primary/secondary election process.

(only if another card is detected in the system)

Reviewer 2: The contradiction is in the wording.

> When a controller card boots up and loads it’s configuration, power-admin-state may be set to POWER_DISABLED. The CONTROLLER_CARD should then power off and never enter into any controller card primary/secondary election process.

rule#4 says to load the config, and if the POWER_DISABLED is detected, it _*never*_ enters the election process.
But then rule#5 redeclares that; the controller card will do the election just fine if the other card is not detected. So it's not never.

> 5. If the chassis has a single controller card and it was configured for config/power-admin-state=POWER_DISABLED, as per Rule#1 above, its state/power-admin-state will stay as POWER_ENABLED given that its state/redundant-role=PRIMARY. However during the next reboot, if the chassis continues to have only one controller-card then it must ignore the configuration of config/power-admin-state=POWER_DISABLED and continue to bootup as PRIMARY. In this situation though, the implementation must send a Syslog message of the severity "Warning" to inform the Operator about the situation. Post reboot, the config/power-admin-state=POWER_DISABLED will stay as is and state/power-admin-state will continue to show as POWER_ENABLED. In this case, state/last-poweroff-time, state/last-poweroff-reason/trigger and state/last-poweroff-reason/details must not be updated

Reviewer 2: This means that the faulty card cannot be disabled if you encounter a failure in communication between the primary and the secondary cards (persistent split brain scenario)

If this is acceptable, I think that should be explicitly mentioned in the text.

Reviewer 2: I didn't ask to add all hypothetical situations :-)

There are a few challenges that I see, and why I asked about this specifically.

In one of the stated objectives, you say

> Operator suspects a network device issue is occurring due to faults on a CONTROLLER_CARD. The operator wants to quickly disable the CONTROLLER_CARD and later, troubleshoot the issue in a maintenance window. The operator also wants to ensure that the card stays disabled even if the device is rebooted.

So, let's say, an operator has a problem where split brain scenario has occurred - he definitely will want to disable the failed card. But now he cannot truly disable it, if the failed card is not able to see the other card, and thinks that it's primary. Because it follows the rules specified in this document (the only card is not disabled immediately, and post reboot has to check if it's the only card and boot successfully if it is)

Thus, I think it is important to call out that the rules specified here explicitly disallow using the `POWER_DISABLED` configuration to mitigate the split-brain issue if it occurs.

And no, I don't agree that
 > Possibility of such a command to power-off a controller-card using configuration do not change how the implementation should otherwise deal with operational challenges like a Split-brain scenario.

This document
a) enforces a particular architecture of NOS boot process, and active-standby selection (which, in my opinion, should not be standardized by OpenConfig)
b) says that power-off rules cannot be used to mitigate the split brain in certain conditions, where another set of rules w.r.t a single card scenario behavior may produce a completely different result.

Author: `So, let's say, an operator has a problem where split brain scenario has occurred - he definitely will want to disable the failed card. But now he cannot truly disable it, if the failed card is not able to see the other card, and thinks that it's primary.`

***Response:** 
- At first, we are also assuming that the operator is able to use push config to the box in a split brain scenario.

`Thus, I think it is important to call out that the rules specified here explicitly disallow using the POWER_DISABLED configuration to mitigate the split-brain issue if it occurs.`

***Response***
I strongly feel that is implicit in the rule already, a Primary controller-card will not be turned off during run time (in a split-brain scenario both cards tend to be PRIMARY and hence none can be turned-off if somehow the operator manages to configure one). Also, if it deems itself as the lone controller-card during reboot, it will continue with the boot process (therefore in the split-brain scenario both cards will continue to boot. However the NOS is expected to have proper safety in place to prevent such a situation at boot). Calling out specific operational failure example in the rules will be repetitive and unproductive. But, this is covered in the `Concerns and possible failure scenarios` section below.

a) enforces a particular architecture of NOS boot process, and active-standby selection (which, in my opinion, should not be standardized by OpenConfig)

**Response**
For this process to work as expected, it is important that the implementation learns what is in the config prior to starting the election.

`b) says that power-off rules cannot be used to mitigate the split brain in certain conditions, where another set of rules w.r.t a single card scenario behavior may produce a completely different result.`

**Response**
Right, that is the expectation and I dont see any ambiguity there. Attached flowchart must help as well. Reiterating the rules in context
- Primary controller card cannot be powered-off, only the secondary-controller card can be powered-off during runtime.  During Split-brain situation when both cards are Primary at run-time, and somehow if the operator can configure one card as disabled, it will stay up because it deems itself as PRIMARY.
- During the boot process, if the controller-card is the only card, then the POWER_DISABLED configuration must be ignored and the boot process should continue. Now if during the boot process there is a split brain situation then the implementation MUST have proper controls in place for protection. Introduction of config/power-admin-state = POWER_DISABLED/POWER_ENABLED is orthogonal to such an expectation for protection from the implementation.

/components/component[name=controller1]/state/redundant-role, PRIMARY
/components/component[name=controller1]/controller-card/config/power-admin-state, POWER_ENABLED
/components/component[name=controller1]/controller-card/state/power-admin-state, POWER_ENABLED
```

2. A controller-card which is in state/redundant-role=SECONDARY and is config/power-admin-state=POWER_DISABLED should remain powered off, even after a reboot. It is possible that after a reboot, both the controller cards are powered-on. However as soon as the configuration is loaded, the system should power-off the subject controller-card.

3. A CONTROLLER_CARD in redundant-role SECONDARY and state/power-admin-state = POWER_DISABLED cannot transition to redundant-role PRIMARY. If the PRIMARY CONTROLLER_CARD goes down, the device will be offline.

4. When a controller card boots up and loads it’s configuration, power-admin-state may be set to POWER_DISABLED. The CONTROLLER_CARD should then power off and never enter into any controller card primary/secondary election process. This also means that an implementation shouldn't start a controller card election process until the configuration is loaded and consumed.

> 5. If the chassis has a single controller card and it was configured for config/power-admin-state=POWER_DISABLED, as per Rule#1 above, its state/power-admin-state will stay as POWER_ENABLED given that its state/redundant-role=PRIMARY. However during the next warm reboot, if the chassis continues to have only one controller-card then it must ignore the configuration of config/power-admin-state=POWER_DISABLED and continue to bootup as PRIMARY. In this situation though, the implementation must send a Syslog message of the severity "Warning" to inform the Operator about the situation. Post reboot, the config/power-admin-state=POWER_DISABLED will stay as is and state/power-admin-state will continue to show as POWER_ENABLED. In this case, if it was a warm reboot then the following must not be updated, state/last-poweroff-time, state/last-poweroff-reason/trigger and state/last-poweroff-reason/details. The same leaves must be updated for cold reboots.

Author: Removed bullet#6. I dont think we need that.

> 5. If the chassis has a single controller card and it was configured for config/power-admin-state=POWER_DISABLED, as per Rule#1 above, its state/power-admin-state will stay as POWER_ENABLED given that its state/redundant-role=PRIMARY. However during the next warm reboot, if the chassis continues to have only one controller-card then it must ignore the configuration of config/power-admin-state=POWER_DISABLED and continue to bootup as PRIMARY. In this situation though, the implementation must send a Syslog message of the severity "Warning" to inform the Operator about the situation. Post reboot, the config/power-admin-state=POWER_DISABLED will stay as is and state/power-admin-state will continue to show as POWER_ENABLED. In this case, if it was a warm reboot then the following must not be updated, state/last-poweroff-time, state/last-poweroff-reason/trigger and state/last-poweroff-reason/details. The same leaves must be updated for cold reboots.

Author: accepted #5

7. Both controller-cards present during the initial boot operation. However, the controller card that is configured as "POWER\_DISABLED" boots up sooner than the other card and takes over the ACTIVE role.

    **Response:**

    The controller cards should never take PRIMARY/SECONDARY roles unless the configuration is fully loaded and processed. Hence, once the configuration is loaded, the card configured to be "POWER\_DISABLED" should shutdown. The only exception is when it is the only controller card in the chassis. In which case, Rule#5 above must be followed.

8. Say we have a situation whereby, A dual controller card device is functional, secondary is powered off by changing power-admin-state to POWER\_DISABLED. Later, the PRIMARY card is removed and the system goes offline. If the device is rebooted (for example via externally removing and reapplying power ), what should happen?

    **Response:**
   
    If during reboot the device continues to have a single controller-card, then Rule#5 above should be followed.

> 9. According to Rule#5, a controller-card must ignore the configuration of `config/power-admin-state=POWER_DISABLED` and must continue with its boot process if it is the only controller-card in the chassis. This also means that, under split-brain situations when the communication between the controller-cards is broken, the option to use config/power-admin-state=POWER_DISABLED for one of the cards may not help post reboot?

Author: `rule https://github.com/openconfig/public/issues/5 only applies to single controller card.`
**Response: ** Right. But the concern is that in a Split-brain scenario both cards think they are the only cards in the box and try to become Primary. More details in the question below.

`Does this mean a scenario where two controller cards are present and both are set to redundancy-state PRIMARY`
**Response: ** `redundant-role` is a state leaf right? IIUC we cant set the `redundant-role` for a controller-card. Having said that, what if the communication between the 2 controller-cards is broken for reasons like software bug, configuration error etc. In that case both cards will consider themselves as the only controller-card in the box and will try to gain Primary role. I will add a definition for Split-brain scenario.

> 9. According to Rule#5, a controller-card must ignore the configuration of `config/power-admin-state=POWER_DISABLED` and must continue with its boot process if it is the only controller-card in the chassis. This also means that, under split-brain situations when the communication between the controller-cards is broken, the option to use config/power-admin-state=POWER_DISABLED for one of the cards may not help post reboot?

Reviewer: Thanks, just looking for the doc to define "split-brain" so it is clear.  Your explanation here makes sense, but it should be in the markdown.  :)

> 9. According to Rule#5, a controller-card must ignore the configuration of `config/power-admin-state=POWER_DISABLED` and must continue with its boot process if it is the only controller-card in the chassis. This also means that, under split-brain situations when the communication between the controller-cards is broken, the option to use config/power-admin-state=POWER_DISABLED for one of the cards may not help post reboot?

Author: done

	**Note:-** If an implementation's architecture do not allow for controlling the order in which the configuration is loaded and the PRIMARY/SECONDARY election process, then this rule can be relaxed as long as the implementation has proper arrangements to power-off the controller-card with `config/power-admin-state=POWER_DISABLED` configuration during reboots (Warm/Cold). 
   

> 5. On boot (cold or warm), if the chassis has a single controller card and it is configured for config/power-admin-state=POWER_DISABLED, it Must continue with the boot process ignoring the configuration. As per Rule#1 above, the controller-card  Must have it's state/power-admin-state as POWER_ENABLED given that a single CONTROLLER_CARD will have it's state/redundant-role=PRIMARY. In this case too, the system must log a message using severity "Warning", to inform the Operator about the situation.

Reviewer: Nit:

```suggestion
5. On boot (cold or warm), if the chassis has a single controller card and it is configured for config/power-admin-state=POWER_DISABLED, it must continue with the boot process ignoring the configuration. As per Rule#1 above, the controller-card  must have it's state/power-admin-state as POWER_ENABLED given that a single CONTROLLER_CARD will have it's state/redundant-role=PRIMARY. In this case too, the system must log a message using severity "Warning", to inform the Operator about the situation.
```
